incubator/hnc/pkg/controllers/object_controller.go

```diff
@@ -177,6 +180,60 @@ func (r *ObjectReconciler) hasCorrectAncestry(ctx context.Context, log logr.Logg
 	return true
 }
 
+func (r *ObjectReconciler) hasOverriddenAncestry(ctx context.Context, log logr.Logger, inst, src *unstructured.Unstructured) bool {
+	if inst == src {
+		return false
+	}
+
+	srcKey := affectedObjectKey(src)
+	sns := r.Forest.Get(src.GetNamespace())
+	if !sns.HasConditionCode(srcKey, api.ObjectDescendantOverridden) {
+		// Early exit if the source doesn't have ObjectDescendantOverridden condition.
+		return false
+	}
+
+	// Find if any ancestor (including itself) has ObjectOverridden condition.
+	ans := r.Forest.Get(inst.GetNamespace())
+	for ans != sns {
+		curKey := replaceKeyNamespace(srcKey, ans.Name())
+		if ans.HasConditionCode(curKey, api.ObjectOverridden) {
+			log.Info("Ancestor modified", "ancestor", curKey)
+			return true
+		}
+		ans = ans.Parent()
+	}
+
+	return false
+}
+
+// isOverridden return false if the instance is not overridden from the source. Otherwise, it will set
+// object conditions in the forest and enqueue affected namespaces to HierarchyReconciler for updating
+// the conditions with the API server, and return true.
+func (r *ObjectReconciler) isOverridden(ctx context.Context, log logr.Logger, inst, src *unstructured.Unstructured) bool {
+	if reflect.DeepEqual(canonical(inst), canonical(src)) {
+		return false
+	}
+
+	ak := affectedObjectKey(inst)
+	sak := affectedObjectKey(src)
+
+	nnm := inst.GetNamespace()
+	ns := r.Forest.Get(nnm)
+	log.Info("Instance modified, setting ObjectOverridden", "affectedKey", ak)
+	msg := fmt.Sprintf("object overridden from source %s", sak)
+	ns.SetCondition(ak, api.ObjectOverridden, msg)
+	r.HierarchyReconciler.enqueueAffected(log, msg, nnm)
+
+	snnm := src.GetNamespace()
+	sns := r.Forest.Get(snnm)
+	log.Info("Descendent modified, setting ObjectDescendentOverridden", "affectedKey", sak)
+	msg = fmt.Sprintf("descendent object %s overridden", ak)
+	sns.SetCondition(sak, api.ObjectDescendantOverridden, msg)
```

Comment:
I think the affected object here should be the _modified_ object, not the source object, because knowing the modified object _and_ the fact that it's in a descendant of the current ns tells us exactly where the current one is.